### PR TITLE
track opam file metadata in dune-project and rewrite opam files

### PR DESCRIPTION
**Pull request by avsm**
This adds support for recording some common metadata in the `dune-project` file that can be used for the purpose of generating packages descriptions.  Initially, common metadata is stored in the toplevel dune-project that can be shared between package managers, and opam-specific metadata is stored within an `opam` field.  This permits extension for direct support of other package managers in the future.

As a reward for putting this metadata in dune-project, dune can now rewrite `opam` files to keep the metadata recorded in dune-project in sync with all the project files. Updating the metadata in dune-project will result in `opam.corrected` files being written and normal dune promotion working.  The current logic aims to leave opam files untouched for fields which are not specified in the dune-project file, and only rewrite ones which are specifically overridden there.

An example dune-project file with the metadata is here:

```
(lang dune 1.10)
(name cohttp)
(source (github mirage/ocaml-cohttp))
(license ISC)
(authors "Anil Madhavapeddy" "Rudi Grinberg")
(opam
  (tags org:mirage org:dune)
  (depends
   (ocaml (>= 4.06.0))
   (cohttp (>= 1.0.0))
  )
  (package
    (name cohttp)
    (synopsis "An OCaml library for HTTP clients and servers")
    (description "A longer description")
    (depends
      (alcotest :with-test)
      (dune (and :build (> 1.5)))
      (foo (and :dev (> 1.5) (< 2.0)))
      (uri (>= 1.9.0))
      (uri (< 2.0.0))
      (fieldslib (> v0.12))
      (fieldslib (< v0.13))))
  (package
    (name cohttp-async)
    (synopsis "HTTP client and server for the Async library")
    (description "A _really_ long description")
    (depends
      (cohttp (>= 1.0.2))
      (conduit-async (>= 1.0.3))
      (async (>= v0.10.0))
      (async (< v0.12))))
)
```

It may be useful to lift the a constraints field up to the toplevel for some packages such as `ocaml`, so that a future PR can use that metadata to generate CI test files automatically.  We can extract that from the opam-specific metadata at present though, and defer that to a future PR.

(this branch is current a WIP, opening this to get more feedback on the approach while hacking with @jonludlam on it)

related to https://github.com/ocaml/dune/issues/1498

### CODE REVIEW

**Comment by jberdine** [comment]
Re feedback on the approach, I wonder if any thought has been given to the related problem of how to keep this metadata in sync with changes made to the opam files in the opam-repository? For example, it has happened that missing dependency version constraints were noticed and added directly to opam-repo, and then when I made a new release the added constraint was lost. Other than manually checking each package's opam file in the repo before each release, I don't know of a work flow to preserve such changes.

**Comment by avsm** [comment]
@jjb that's a really good question, and one I've also observed but not thought of a solution to yet.  It's not quite clear to me whose responsibility (opam tooling _vs_ dune tooling) it should be.  One solution is to put it in the forthcoming duniverse/dune-get tool (https://github.com/avsm/duniverse) which acts as a bridge between both tools.  Your usecase is now on my list as something that needs a solution.

**Comment by jjb** [comment]
@avsm i think you meant to tag @jberdine :) although this makes me dream of working with ocaml again…

**Comment by avsm** [comment]
Woops, I did indeed, sorry! Although if it subliminally encourages you to return to OCaml, then this PR has done its job well :-)

**Comment by rauanmayemir** [comment]
Could there be a way to indicate the libraries that a package provides as well? If dune project could set opam ‘constraints’ for a package, maybe there could be a soft way to have `(libraries)` set based on that information, too? It’s a bit circular: we’d want to know what libraries a package provides so that setting a constraint on that package could gather underlying libraries information. It’s not dune’s job to hold devs’ hand, but it sounds like a perfect job for some future dune-lsp tooling.

The common confusion for beginners is the package->library hierarchy and especially the naming quirks. ‘Admitting’ and protocoling the hierarchy could remove most of the friction.

File: dune.opam [hunk 1]
```diff
@@ -11,7 +11,6 @@ depends: [
   "base-threads"
 ]
 build: [
-  # opam 2 sets OPAM_SWITCH_PREFIX, so we don't need a hardcoded path
```
Comment (rgrinberg) [inline dune.opam]: Why was this comment removed?

**Comment by rgrinberg** [comment]
@jonludlam I've cleaned up the PR and made the opam file diffing work when an extension is enabled in the dune-project file: `(generate_opam_file)`.

One last thing that I'd like to see before the PR is merged is the removal of all pp/sexp functions introduced in favor of `to_dyn` functions.

File: dune.opam (hunk 1 above)
Comment (avsm) [reply dune.opam]: Hm was this removed due to the rewriting of the opam file removing comments. We should test that specifically to make sure comments are preserved

**Comment by rgrinberg** [comment]
I've changed the status of this PR to be ready for review. One thing that I'd like a second opinion on is the name of this option in the dune-project file. Currently, it's `(generate_opam_file)`, but perhaps it should be `(lint_opam_file)`, or `(correct_opam_file)`?

**Comment by rgrinberg** [comment]
@jonludlam I got rid of blang_decode. I'm not sure why it was split this way, but it doesn't seem necessary anymore.

**Comment by avsm** [comment]
@rgrinberg wrote:
> One thing that I'd like a second opinion on is the name of this option in the dune-project file. Currently, it's `(generate_opam_file)`

Why do we need this at all?  Just the presence of the `(opam` clause in `dune-project` could activate this feature, and it could be explicitly disabled with a non-default option within that clause in the future. But if it causes problems, just removing the `(opam` clause should be sufficient.

**Comment by rgrinberg** [comment]
Yeah, I think that's acceptable as well. This would be a breaking change technically however. At least for people who have used this opam field before it did anything.

**Comment by jonludlam** [comment]
@rgrinberg Thanks very much! A few comments - I'm still leaning in the direction of executing this on `runtest` rather than `install` - mostly because I `make test` more often than I `make install`, and for this to be useful to duniverse it needs to be correct all the time. The blang_decode this was because at some point during development of this I got a circular dependency, but I'm very happy it's not necessary any more. The comments in the opam files are a bit annoying, but the current parser simply thows them away.

I've got some improvements to the docs generation that depends on this, but I'll make that in a separate PR when this is merged.

**Comment by rgrinberg** [comment]
>  mostly because I make test more often than I make install, and for this to be useful to duniverse it needs to be correct all the timee

Then I suppose there's no need to choose. We could make it a dependency of both aliases.

> The comments in the opam files are a bit annoying, but the current parser simply thows them away.

Perhaps we should not do any correction if any comments are encountered? I don't think we can merge the code in its current state if it's going to drop user comments.

**Comment by jonludlam** [comment]
There is a 'format preserving printer' in v2.0.0 of opam-file-format which attempts to preserve comments - dune currently has 2.0.0~beta which predates it. Let me try upgrading that and seeing if we can use that to fix the opam comment issue.

**Comment by jonludlam** [comment]
That seems to do the trick. Just need to enable the rewriting on existence of (opam) now and hook it up to the runtest target as well as install.

**Comment by avsm** [comment]
@rauanmayemir I'm still thinking about your comment about merging libraries.  It's fairly intertwined with opam's "special" status in dune as the unit of `(package)` definitions.  I'll bring this up at the next dune developer meeting for a broader discussion to see how we can remove any duplication with this PR.

**Comment by avsm** [comment]
@jonludlam phew! That would have been quite the blocker for automated rewriting of opam files otherwise.

File: test/blackbox-tests/test-cases/dune-project-meta/proj [hunk 2]
```diff
@@ -0,0 +1,31 @@
+(lang dune 1.10)
+(name cohttp)
+(source (github mirage/ocaml-cohttp))
```
Comment (avsm) [inline test/blackbox-tests/test-cases/dune-project-meta/proj]: Just a thought here: we could shorten this to the common case of

```
(source mirage/ocaml-cohttp)
```

to make the `(github mirage/ocaml-cohttp)` the default.  Can the dune parser do the magic default if just an Atom is specified instead of a Sexp field?

Comment (rgrinberg) [reply test/blackbox-tests/test-cases/dune-project-meta/proj]: Yes it can. Is it really necessary however? I like github but perhaps we should keep support open for other hosting providers

Comment (avsm) [reply test/blackbox-tests/test-cases/dune-project-meta/proj]: Not suggesting removing support for other providers; just making the common case shorter syntax default to github (as e.g. the brew CLI does). I’m also fine with leaving as is

File: test/blackbox-tests/test-cases/dune-project-meta/proj [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  )
+  (package
+    (name cohttp)
+    (synopsis "An OCaml library for HTTP clients and servers")
+    (description "A longer description")
+    (constraints
+      (>= uri 1.9.0)
+      (< uri 2.0.0)
+      (< fieldslib v0.12)
+      (< fieldslib v0.12)))
```
Comment (avsm) [inline test/blackbox-tests/test-cases/dune-project-meta/proj]: this is a duplicate - needs to be > something?

**Comment by jonludlam** [comment]
Just had a chat with @avsm - as it stands this PR does nothing with the constraints in dune-project,  let's fix that so that the constraints are fixed in the opam files before we merge this.

**Comment by jonludlam** [comment]
I've changed the format of `constraints` and renamed it `depends`. It now more closely matches opam's, but is a bit more narrow in what it can describe.

It now looks like this:

```
(depends
      (alcotest :with-test)
      (dune (and :build (> 1.5)))
      (foo (and :dev (> 1.5) (< 2.0)))
      (uri (>= 1.9.0))
      (uri (< 2.0.0))
      (fieldslib (> v0.12))
      (fieldslib (< v0.13))))
```

**Comment by avsm** [comment]
@jonludlam and I went through quite a few syntax variations in order to come up with the above, and we think it's pretty robust to all the dune-universe usecases we could spot.  A re-review of it would be appreciated before this is merged.

@jberdine I've created a duniverse issue to track your idea separately https://github.com/avsm/duniverse/issues/16

**Comment by ejgallego** [comment]
Would be awesome if the syntax would use less parenthesis:
```
(depends
      (alcotest :with-test)
      (dune :build and > 1.5)
      (foo :dev and > 1.5 and < 2.0)
      (uri >= 1.9.0)
      (uri < 2.0.0)
      (fieldslib > v0.12)
      (fieldslib < v0.13))
```

**Comment by rgrinberg** [comment]
IMO, the lispy syntax is more clear. However, @jonludlam, @avsm: I'm wondering if we should make part of this feature a bit less opam specific. Add a toplevel field called `package` or something. Things like `tags` should indeed stay under opam however. I'm thinking that once we create a little dune library to query the dune-project file, this information could be useful in contexts outside of opam.

**Comment by rgrinberg** [comment]
@jonludlam one issue with enabling the rules only if the opam stanza is present is that we no miss out on the validation of the other fields in some situations. It's a bit strange that we aren't verifying license, authors, dev-repo just because we haven't listed the opam dependencies.

**Comment by rgrinberg** [comment]
Why did we add support for `!=` for inequality? It seems like it will add just a bit of entropy for not much gain. How about we stick to one syntax and just have dune emit a good error message if the other operator is used.

**Comment by rgrinberg** [comment]
I've added support for the conflicts field and have also populated all the metadata in dune's project file. It's a good idea to dogfood this stuff to iron out the kinks.

**Comment by rgrinberg** [comment]
There's a few more fields that dune can be smarter about: `maintainer`, `homepage`, `bug-reports`, `build`. There's already enough information to fill in those fields for the common case, we just need to allow the user to override these fields when they need something custom.

**Comment by rgrinberg** [comment]
One more point:

Should dune be a little smarter when there are no `package` fields? Currently, if the package field is absent, it will not suggest any corrections. Instead, perhaps it should just assume that there's a single package and it's equivalent to the dune project's name?

**Comment by avsm** [comment]
> Should dune be a little smarter when there are no package fields? Currently, if the package field is absent, it will not suggest any corrections. Instead, perhaps it should just assume that there's a single package and it's equivalent to the dune project's name?

Yes this sounds sensible to me.  I didn't realised it didn't do this already!

File: dune-project [hunk 4]
```diff
@@ -1,4 +1,38 @@
-(lang dune 1.8)
+(lang dune 1.10)
 (name dune)
 
-(implicit_transitive_deps false)
\ No newline at end of file
+(implicit_transitive_deps false)
+
+(license MIT)
+(authors "Jane Street Group, LLC <opensource@janestreet.com>")
```
Comment (avsm) [inline dune-project]: The authors list is an interesting one. I think we should encourage the listing of individual contributions as well in order to keep this accurate. What do you think about having a general `author` schema that could be

```
(maintainers @diml @rgrinberg)
(authors "Jane Street Group <opensource@janestreet>" (gitlab @avsm))
```

The idea here is that maintainers are always authors, and that we are flexible about how to identify someone. A string starting with @ is assumed to be github by default.

(this can probably be deferred for a future PR, but I just thought about it now!)

Comment (diml) [reply dune-project]: I always wonder what to put in this field BTW. For instance, should it be every single person who submitted a PR? The "Insights" tab is much more precise in this regard.

Comment (avsm) [reply dune-project]: It's a value judgement of "significant contributions".  So a typo fix doesnt cut it, but a reasonably small feature that adds some code does.

Comment (diml) [reply dune-project]: Alright. BTW, are maintainers really always authors? For instance, there are debian maintainers of the dune package that are not authors.

Comment (avsm) [reply dune-project]: That's true for packages (since its a metadata overlay), but not (I think) ever true for source repos of the sort dune manages.  It does mean we'll need a package-specific maintainer field in the package manager areas.  So in the `(opam)` clause we could also allow a maintainer field in the future to override the global one in this PR.

I'm thinking that in the medium term dune or extensions can make smarter use of the global fields in order to improve error messages.  For instance a build failure could be automatically uploaded to a github issue in the right repo, or dune could even search online for messages from the `maintainers` who answer FAQs.

File: src/opam_create.ml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    let* opam = Dune_project.opam project in
+    let+ pkg = Dune_project.Opam.find opam package_name in
+    (opam, pkg))
+    correct_specific >>>
+  opt (Dune_project.license project) (set_string "license") >>>
+  list (Dune_project.authors project) (set_list "authors" mkstring) >>>
+  opt (Dune_project.version project) (set_string "version") >>>
+  opt (Option.map ~f:(Format.asprintf "%a" Dune_project.Source_kind.pp)
+         (Dune_project.source project)) (set_string "dev-repo") >>>
+  set_string "opam-version" "2.0" >>>
```
Comment (avsm) [inline src/opam_create.ml]: we could also add a `x-generated-by: dune-<version>` field here into the opam file so that we can track adoption of this feature.

File: dune-project (hunk 4 above)
Comment (diml) [reply dune-project]: Ok. That makes sense. So I guess the way to describe these fields is:
- maintainers: the people currently maintaining the project
- authors: everybody who has made significant contributions to the project

**Comment by rgrinberg** [comment]
To make this feature more useful, we should consider adding the following checks:

* dune can already discover the local package dependencies of a package, we could let the user know if such a package is missing from the list in the dune-project file.

* Dune can consider all the library dependencies of all executables and libraries that belong to a library. It would then turn this into a package list and make sure that the user hasn't omitted any of the package dependencies inferred this way. Yes, this isn't perfect as findlib and opam package names do not match for some non-dune packages. As a workaround, we can add a way for the user to associate findlib to package names. This will not capture all dependencies either, nevertheless, it will be nice for the common case.

This last suggestion is a bit unrelated, but i've been meaning to bring it up before:

Currently, dune supports tests that don't belong to any package. That seems like a bit of a hidden pitfall to me because opam is executing the test suites per package. This results in the same tests being run multiple times. I'm wondering if we should warn users about such dangling tests, if this is desired, we can have them mark such tests explicitly: `(package (dangling))`.

**Comment by avsm** [comment]
All three of those suggestions seem like excellent ideas. I wonder if it's better to merge this PR and then do them separately, since they cover some semantic changes to dune files.

**Comment by diml** [comment]
One second, where is doc? :) We don't need to describe everything in detail, but the feature should at least be mentioned in the `doc/` directory.

**Comment by rgrinberg** [comment]
I would like to squash all the commits as well, as the history isn't making much sense.

**Comment by rgrinberg** [comment]
I will add some brief docs and merge as this feature isn't complete yet.

@avsm we can merge this PR now, but I'd caution us against releasing features piecemeal. The issue is that it puts a versioning burden down the line. For example, if we add a `depopts` to dune-project, we would need to make it >= 1.11. That complicates things for the user and complicates our codebase with this versioning overhead, so it should be avoided if possible.

**Comment by rgrinberg** [comment]
I've jotted down some documentation in a new opam section. I think we need such a section because there's quite a few opam specific things we can document. The release workflow with dune-release is one glaring example of something that isn't so easy to lookup.

An alternative layout would be for this information to live in the dune project section instead, and the opam section can simply link to the appropriate dune-project section.

File: doc/opam.rst [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+The fields that dune uses for this purpose are:
+
+- ``(license <name>)`` - Specified the license of the project
+
+- ``(authors <authors>)`` - A list of authors
+
+- ``(source <source>)`` - where the source is specified two ways:
+  ``(github <user/repo>)`` or ``(uri <uri>)``
+
+To enable dune suggesting corrections to the opam stanza, the user must specify
```
Comment (diml) [inline doc/opam.rst]:
```suggestion
To enable dune suggesting corrections to the opam file, the user must specify
```

**Comment by diml** [comment]
Having a dedicated opam section in the manual seems good. Though, we should link it from the `dune-project` doc.

BTW, in the past, people have asked for a way to declare packages without writing `<package>.opam` files because they don't target opam. The natural answer would be to declare the list of packages in the `dune-project` file. However, if one doesn't target opam then it does feel right to write `(opam (package ...))` to declare the list of packages. Instead, we could declare packages at the toplevel and write the opam specific fields inside it:

```
(lang dune 1.10)
(name foo)

(package
 (name foo)
 (opam <opam-specific-fields>))

(package
 (name bar)
 (opam <opam-specific-fields>))
```

Then someone who is not targetting opam would simply omit the `opam` fields. Sorry I only bring this up now, I didn't have the time to look at this PR before and only noticed this while looking at the documentation.

**Comment by andreypopp** [comment]
> BTW, in the past, people have asked for a way to declare packages without writing <package>.opam files because they don't target opam. The natural answer would be to declare the list of packages in the dune-project file.

I was about to comment on this. Thank you @diml for bringing this up.

It would be great to have a way to declare a package with dune without mentioning opam specifically as with esy we might not have an opam package in the first place (I think it might be confusing for novice esy users).

I'd also suggest not to include the cascading behaviour for `depends` and `conflicts` stanzas (an ability to specify those for all packages and for concrete pakcages). Not sure if it's worth it. I like that you can look at a a single package and see all its dependencies declared in a single place without the need to "merge" formulas. Maybe if boilerplate has become a problem in dune configs then it can be solved by some generic mechanism (named values)?

**Comment by avsm** [comment]
The `(package)` stanza makes sense to me.  It would be really good to have a non-opam field there to ensure we haven't missed something. @andreypopp, would you have an example of what an esy file generated by these package stanzas might look like?

I also think the `(package)` field makes it possible to tighten up the slightly weird "dangling tests" semantics that @rgrinberg brought up earlier.  We should ensure that everything is tied to exactly one package (so it isn't run multiple times), with a default package of the project name if there are no `(package)` stanzas.

**Comment by rgrinberg** [comment]
> I'd also suggest not to include the cascading behaviour for depends and conflicts stanzas (an ability to specify those for all packages and for concrete pakcages). Not sure if it's worth it

@avsm what do you think? Having statically evaluated variables would indeed solve this problem and other similar ones. The `depends` boilerplate will be partially alleviated once dune is able to partially infer package deps.

**Comment by avsm** [comment]
I'm fine with the simpler non-cascading behaviour to start with, even if it's more verbose.  There simply isn't that much overlap at the moment since we only have the opam backend.

**Comment by avsm** [comment]
Thanks for the squash commit @rgrinberg -- I think it also needs a `Co-authored-by: Jon Ludlam <jon@recoil.org>` as lots of the intermediate commits were his. This seems to be a cool new feature in GH: https://help.github.com/en/articles/creating-a-commit-with-multiple-authors

**Comment by andreypopp** [comment]
@avsm I don't think we want to generate esy metadata out of dune files, instead we can support dune-project files in esy directly.

I'm mainly interested in not having `*.opam` files at the root of projects managed with esy and then not having to specify opam metadata for non-opam packages (hence I'm asking for `(package ...)` stanza which is not specific to opam).

Ideally though with esy I'd like not to duplicate package name in multiple places (being it `NAME.opam` file or mentioning `(package NAME)` in `dune-project`) and Instead configuring dune with `DUNE_PACKAGE_NAME` or similar.

**Comment by avsm** [comment]
Sounds good. Do you have an example of a dune-managed package that doesn't have a corresponding opam package, anywhere?  So far it seems that all dune packages do map to opam packages.

(I'm primarily aware that we don't have any non-opam tests at the moment, so some concrete example of using the non-opam package would be useful so we can ensure we do the right thing)

**Comment by andreypopp** [comment]
> Do you have an example of a dune-managed package that doesn't have a corresponding opam package, anywhere? So far it seems that all dune packages do map to opam packages.

Yes! Having `*.opam` files is the requirement to use dune right now. This is why we are forced to create them even if they are completely empty — https://github.com/esy-ocaml/hello-reason/blob/master/hello-reason.opam. This is confusing for newcomers and I hope `(package ...)` stanza will make it less so. Though ideally, as I said, we don't have to duplicate package name in multiple places.

**Comment by avsm** [comment]
Got it -- so it's for the very important getting-started usecase, but nothing else more complex that we're missing.  It will be nice to have a dune project description with just the `dune-project` and `dune` file to start with.

**Comment by jonludlam** [comment]
We could make it such that if you don't specify the package name it inherits the top-level name (and also ensure that no two packages have the same name)

```
(lang dune 1.10)
(name dune)

(package
  (opam
    (tags foo bar baz)))
```

Implicitly names that package 'dune'.

We could also change the check for a .opam file to additionally check for a (package) stanza in dune-project with the correct name - does that sound right?

**Comment by rgrinberg** [comment]
> Thanks for the squash commit @rgrinberg -- I think it also needs a Co-authored-by: Jon Ludlam <jon@recoil.org> as lots of the intermediate commits were his. This seems to be a cool new feature in GH: help.github.com/en/articles/creating-a-commit-with-multiple-authors

Thanks, will do so.

> We could make it such that if you don't specify the package name it inherits the top-level name (and also ensure that no two packages have the same name)

Seems good to me.

I will merge this PR as is, that should make it easier for all of us to contribute.

The agenda:

* Allow dune to define packages purely in the dune-project file
* Make the `package` stanza independent of opam
* Remove shared conflicts/depends
* Assume there's a single package named after the dune-project if there are no opam files.

@jonludlam Currently, the correction rules do no work if the opam file doesn't exist. I think dune should promote the opam file it assembled from the dune-project file.
